File: leetcode/1339.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def prefix_sum(self, root: TreeNode):
        if root is not None:
            left = self.prefix_sum(root.left)
            right = self.prefix_sum(root.right)
            return root.val + left + right
        return 0

    def find_closer_subtree(self, root: TreeNode):
        if root is not None:
            left = self.find_closer_subtree(root.left)
            right = self.find_closer_subtree(root.right)
            my_total = root.val + left[2] + right[2]
            my_current_diff = abs(my_total - self.target)
            return (root, min([left[1], my_current_diff, right[1]]), my_total)
        else:
            return (None, float('inf'), 0)
        
    def maxProduct(self, root: TreeNode) -> int:
        total = self.prefix_sum(root)
        self.target = total / 2
        result = self.find_closer_subtree(root)
        return (int)((self.target + result[1]) * (self.target - result[1])) % (10**9 + 7);

# if __name__ == "__main__":
#     app = Solution()
#     print(app.maxProduct(TreeNode(10, TreeNode(11), None)))

File: leetcode/1339.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sum(root: TreeNode | null): number {
    if (root) {
        let left = sum(root.left);
        let right = sum(root.right)
        return (root.val + left + right);
    } else {
        return 0;
    }
};

function findCloserSubtree(root: TreeNode | null, target: number) : [TreeNode | null, number, number] {
    if (root) {
        let left_best = findCloserSubtree(root.left, target);
        let right_best = findCloserSubtree(root.right, target);
        let my_sum = root.val + left_best[2] + right_best[2];
        //change the values to send the best to return
        left_best[2] = my_sum;
        right_best[2] = my_sum;
        let my_current : [TreeNode | null, number, number] = [root, Math.abs(my_sum - target), my_sum];
        if (left_best[1] <= right_best[1] && left_best[1] <= my_current[1]) {
            return left_best;
        } else if (right_best[1] <= left_best[1] && right_best[1] <= my_current[1]) {
            return right_best;
        } else {
            return my_current;
        }
    } else {
        return [null, Number.MAX_SAFE_INTEGER, 0];
    }
};

// achar subarvore com valor mais próximo da metede do valor total da arvore
function maxProduct(root: TreeNode | null): number {
    let total = sum(root);
    let half = total / 2;
    let closer = findCloserSubtree(root, half);    
    let total_closer = sum(closer[0]);
    let res = (total_closer * (total - total_closer)) % (10**9 + 7);
    return res;
};
